Port storytext to use ICU BreakIterator

Fahad-Alsaidi opened the pull request:
By this commit most used storytext functions ported to use ICU BreakIterator. However, there are more to port but not used anywhere.

For testing, use check spelling  from extras menu.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -198,21 +199,22 @@ void StoryText::moveCursorLeft(bool isGrapheme)
+			return;
+
+		int pos = cursorPosition();
+		it->setText(plainText().utf16());
```
Since we always call `setText()` after getting the iterator, I’d move to the `get*Iterator()` method and pass `painText()` to it. This also makes sure no use site misses setting the text.

Fahad-Alsaidi replied:
I tried this but cases a lot of problem. Iterator starts mix text and never work with CJK. I think the problem that we use static vars which is unsafe thread. So better leave it as it unless there is somebody willing to dig ICU source code to find the problem.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -198,21 +199,22 @@ void StoryText::moveCursorLeft(bool isGrapheme)
-			boundaryFinder.setPosition(cursorPosition() - 1);
-			for (pos = -1; !boundaryFinder.isAtBoundary(); pos--)
-				boundaryFinder.setPosition(boundaryFinder.position() - 1);
+			pos = cursorPosition();
```
This is redundant, you already set `pos = cursorPosition()` before the if block.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -252,21 +254,22 @@ void StoryText::moveCursorRight(bool isGrapheme)
-			boundaryFinder.setPosition(cursorPosition() - 1);
-			for (pos = -1; !boundaryFinder.isAtBoundary(); pos--)
-				boundaryFinder.setPosition(boundaryFinder.position() - 1);
+			pos = cursorPosition();
```
Same here.

khaledhosny replied:
After removing this line, you can remove the curly brackets around the blocks as they will be redundant.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1516,20 +1519,14 @@ int StoryText::firstWord()
+
+	it->setText(plainText().utf16());
+	int len = length();
+	int nxt = it->following(pos);
```
Just make is `pos = …` and return pos, instead of the ugly variable name.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1516,20 +1519,14 @@ int StoryText::firstWord()
-		return len;
+
+	it->setText(plainText().utf16());
+	int len = length();
```
This variable is unused, are you sure the new code matches the behaviour of the old one?

Fahad-Alsaidi replied:
Yes I sure. You can test by using Spelling check from extras menu.

khaledhosny replied:
Remove the variable then.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1540,49 +1537,53 @@ int StoryText::prevWord(int pos)
 
-int StoryText::endOfWord(int pos) const
+int StoryText::endOfWord(int pos)
 {
```
Why was the `const` removed?

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1516,20 +1519,14 @@ int StoryText::firstWord()
+	return nxt;
 }
 
 int StoryText::prevWord(int pos)
```
This should be ported as well, to match the method above.

Fahad-Alsaidi replied:
This function not used anywhere. So no point to port it.

khaledhosny replied:
Either port it for completeness sake, since it might get be used in the future and it will be wrong then, or completely remove it. I prefer keeping it for API parity, it does not hurt much.

khaledhosny commented on scribus/text/storytext.h:
```diff
@@ -274,12 +274,15 @@ public slots:
+	static BreakIterator* m_sentenceIterator;
+	BreakIterator* getGraphemeIterator();
 	BreakIterator* getWordIterator();
+	BreakIterator* getSentenceIterator();
```
These methods should be static as well, I missed that for `getWordIterator()`.

Fahad-Alsaidi commented:
Updated. Please notice that I return endOfWord back the const and make all similar functions const. Also I port prevWord & prevSentence to ICU even they are not used anywhere. I also modified to adopt the new change.

khaledhosny commented on scribus/plugins/tools/hunspellcheck/hunspellpluginimpl.cpp:
```diff
@@ -181,7 +181,7 @@ bool HunspellPluginImpl::parseTextFrame(StoryText *iText)
 			wordsToCorrect.append(wf);
 		}
-		currPos = iText->nextWord(wordStart);
+		currPos = iText->endOfWord(wordStart);
```
Why was this changed?

Fahad-Alsaidi replied:
because nextWord will give us word+space and we don't want it.

khaledhosny replied:
But then the new `nextWord()` is wrong, it should match exactly the behaviour of the old one.

Fahad-Alsaidi replied:
In that case I will make nextWord call endOfWord because they will be same just duplicated code. Is that right?

khaledhosny replied:
If this is the only place this was calling `nextWord()`, then keep this change and just kill `nextWord()` and `prevWord()` since both will be unused now.

Fahad-Alsaidi replied:
As we agreed on https://github.com/HOST-Oman/scribus/issues/166 I will not remove any unused code.

khaledhosny replied:
Yes, I was under the assumption that we are not touching that unused code at all, but if we need to modify it because it is broken then we are better of just killing it.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -167,7 +168,7 @@ int StoryText::cursorPosition() const
 }
 
-void StoryText::setCursorPosition(int pos, bool relative)
+void StoryText::setCursorPosition(int pos, bool relative) const
```
This changes the cursor position, so it shouldn’t be const. Only methods that changes nothing in the class should be const.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
+	pos = it->next();
```
Why is the call to `next()` here?

Fahad-Alsaidi replied:
to get the begging of the next word after the space.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
+		return pos;
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
```
This should `preceding()`, shouldn’t it?

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
+	int next = it->next();
```
Just make it `pos = it->next()` and return `pos`, no need for a new variable. Similar to code above.

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1514,75 +1508,76 @@ int StoryText::firstWord()
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
+	int prev = it->previous();
```
Same here.

Fahad-Alsaidi commented:
updated!

khaledhosny commented on scribus/text/storytext.cpp:
```diff
@@ -1579,9 +1570,13 @@ int StoryText::prevSentence(int pos)
+		return pos;
+
+	it->setText(plainText().utf16());
+	pos = it->following(pos);
```
This should be `preceding()` as well.

Fahad-Alsaidi commented:
updated!

andreas-vox commented:
Question: what are the advantages of the ICU break iterator over Qt's QTextBoundaryFinder?

khaledhosny commented:
@Fahad-Alsaidi reported that the word iterator was not working for some languages, also ICU BreakIterator is more widely used and tested, and gets updated regularly for new Unicode versions. Qt itself is going to use ICU at some point.

Fahad-Alsaidi commented:
@andreas-vox  For your reference please see these bugs:
1- [QTBUG-53034 ](https://bugreports.qt.io/browse/QTBUG-53034) QTextBoundaryFinder::Word doesn't work in CJK
2- [QTBUG-29930](https://bugreports.qt.io/browse/QTBUG-29930) Unicode script itemization code needs to be re-worked
3- [QTBUG-42078](https://bugreports.qt.io/browse/QTBUG-42078) Support Lao Word Boundary Detection and Line Breaking
4- [QTBUG-33763](https://bugreports.qt.io/browse/QTBUG-33763) Update Unicode database and algorithms up to 8.0

andreas-vox commented:
Thanks for the links and your good work on CTL!

So currently Harfbuzz + ICU are state of the art?

> Am 08.05.2016 um 06:51 schrieb Fahad-Alsaidi notifications@github.com:
> 
> @andreas-vox https://github.com/andreas-vox For your reference please see these bugs:
> 1- QTBUG-53034  https://bugreports.qt.io/browse/QTBUG-53034QTextBoundaryFinder::Word doesn't work in CJK
> 2- QTBUG-29930 https://bugreports.qt.io/browse/QTBUG-29930 Unicode script itemization code needs to be re-worked
> 3- QTBUG-42078 https://bugreports.qt.io/browse/QTBUG-42078 Support Lao Word Boundary Detection and Line Breaking
> 4- QTBUG-33763 https://bugreports.qt.io/browse/QTBUG-33763 Update Unicode database and algorithms up to 8.0
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub https://github.com/HOST-Oman/scribus/pull/172#issuecomment-217691205

khaledhosny commented:
Yes AFAIK, that is pretty much what everyone is using or migrating to (Firefox, Chrome, LibreOffice, Java etc.) Even InDesign uses parts of ICU (but not HarfBuzz yet AFAIK, we have got a big head start over them in this area 😃).
